File: implementations/java/src/main/java/com/webweavex/Chunker.java
package com.webweavex;

import java.util.*;

public class Chunker {
    private int size = 500;
    private int overlap = 50;
    
    public Chunker() {}
    
    public List<Map<String, Object>> chunk(String text) {
        List<Map<String, Object>> chunks = new ArrayList<>();
        if (text == null || text.isEmpty()) return chunks;
        
        int start = 0;
        int index = 0;
        
        while (start < text.length()) {
            int end = start + size;
            
            if (end < text.length()) {
                end = findWordBoundary(text, end);
            }
            
            String chunkText = text.substring(start, Math.min(end, text.length()));
            if (!chunkText.trim().isEmpty()) {
                Map<String, Object> chunk = new LinkedHashMap<>();
                chunk.put("text", chunkText);
                chunk.put("index", index);
                chunk.put("start", start);
                chunk.put("end", end);
                chunks.add(chunk);
                index++;
            }
            
            start = end - overlap;
            if (start < 0) start = 0;
        }
        
        return chunks;
    }
    
    private int findWordBoundary(String text, int position) {
        if (position >= text.length()) return position;
        
        for (int i = position; i > Math.max(0, position - 50); i--) {
            char c = text.charAt(i);
            if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
                return i;
            }
        }
        
        return position;
    }
}

File: implementations/kotlin/src/main/kotlin/com/webweavex/Chunker.kt
package com.webweavex

data class Chunk(
    val text: String,
    val index: Int,
    val start: Int,
    val end: Int
) {
    fun toMap(): Map<String, Any> = mapOf("text" to text, "index" to index, "start" to start, "end" to end)
}

class Chunker {
    private val size = Config.CHUNK_SIZE
    private val overlap = Config.CHUNK_OVERLAP

    fun chunk(text: String): List<Chunk> {
        if (text.isEmpty()) return emptyList()

        val chunks = mutableListOf<Chunk>()
        var start = 0
        var index = 0

        while (start < text.length) {
            var end = start + size

            if (end < text.length) {
                end = findWordBoundary(text, end)
            }

            val chunkText = text.substring(start, minOf(end, text.length))
            if (chunkText.isNotBlank()) {
                chunks.add(Chunk(chunkText, index, start, end))
                index++
            }

            start = end - overlap
            if (start < 0) start = 0
        }

        return chunks
    }

    private fun findWordBoundary(text: String, position: Int): Int {
        if (position >= text.length) return position

        for (i in position downTo maxOf(0, position - 50)) {
            val c = text[i]
            if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
                return i
            }
        }

        return position
    }
}
